### ISSUE TITLE
Partition tests into groups using id  =~

### ISSUE THREAD

**Issue by gilescope** [nunit/nunit#1038]
Currently using regex (=~) with id leads to an error. For non-thread-safe codebases where test fixture parallelization will never be an option, to keep the test duration down we either have to keep splitting and re-splitting the tests into smaller dlls, or manually marking them up into different categories.

Ideally we'd like to use the test selection language to 'modulus' the tests into groups. For example all tests with an odd id get run on one machine and all tests with an even id get run on another machine. 

I.e. we give NUnit all the test assemblies and it runs half of the tests in each of them. This would enable more parallelization as there would be more nunit-agent processes running than if half the dlls were run on one machine and half on another.

At the moment it can achieved by partitioning the tests into groups via regexing on the test name, but that's likely to be more lumpy than using id. 

Very keen to hear suggestions on the 'right' way to partition tests without introducing additional dlls. I'm sure this will not be an uncommon use case.

**Comment by CharliePoole** [comment]
Note that the "error" from using a regex on the test id is deliberate. We only support equality selection on ids. An earlier version of the documentation made that clear but it no longer does. I'll update the docs.

The reason for allowing only selection on equality is that the user has no information about the form of the test id. In fact, even selection on equality only makes sense for a program like a gui, which locates all the tests and maintains a dictionary of their ids. I'll try to figure out a way to explain this in the documentation - or perhaps remove id from the docs entirely.

There are three general ways that people use to partition tests for separate runs: assembly, namespace and category. For my teams, I have always advocated using them in exactly that order. So if you prefer not to have multiple assemblies, I'd suggest you next look at the namespaces. Generally, tests in the same namespace tend to use the same resources. You'll have to analyze your tests to see if that's true for you. As a last resort, you can assign execution categories to your tests and make separate runs that use different categories.

NUnit's design, as it now stands, is intended to encourage you to use multiple threads in order to promote throughput. This takes some work in order to identify which fixtures are able to run in parallel, but the path you are following of identifying tests randomly will take quite a bit of work too. Running a test in a separate process will not prevent conflicts between fixtures, it will only reduce them.

In your situation, I would start by identifying individual fixtures that can run in parallel, marking them with the ParallelizableAttribute. Unless they are using external resources, there should be little conflict across fixtures for well-designed tests.

**Comment by CharliePoole** [comment]
I'm closing this as "won't fix" for the specific solution you have asked for: partitioning tests by id. This has too many problems, since the id is intended to be an opaque string and may vary in nature from one driver to another.

There may be other solutions to the problem that we could consider as a future enhancement, but I strongly suggest you try using the multi-thread fixture first. Post on the discussion list if you need help in making this work.

**Comment by CharliePoole** [comment]
The documentation has been updated to clarify that id is not intended for use by users.
